### ISSUE TITLE
Cannot pair with passkey ...

### ISSUE THREAD

**Issue by rene-becker-setec** [ThomasGerstenberg/blatann#115]
Hi,

This is a great module for driving Nordics connectivity FW on my nRF52 DK.

It works great if I don't make use of Bluetooth security, but I am having trouble pairing once I do enable security.
I need to pair with a 6 digit passkey using the 'keyboard only' method.

I get to the point where the pass-key entry callback is called (the one registered via `peer.security.on_passkey_required.register(...)`, but immediately after I get an authentication failure.

Here's an example log

```
[2022-05-30 13:58:11,089] [MainThread] [blatann.gap.default_bond_db.load:33] [INFO]: Loaded Bond DB '/home/rene/.blatann/bonding_db.pkl'
[2022-05-30 13:58:11,089] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] open()
[2022-05-30 13:58:12,293] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_conn_configure(<blatann.nrf.nrf_types.config.BleConnConfig object at 0x7fb3903f64c0>,)
[2022-05-30 13:58:12,298] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_enable(<blatann.nrf.nrf_types.config.BleEnableConfig object at 0x7fb3903f6460>,)
[2022-05-30 13:58:12,304] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_device_name_set(b'nRF5x',)
[2022-05-30 13:58:12,306] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_appearance_set(<Appearance.unknown: 0>,)
[2022-05-30 13:58:12,309] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_connect(PeerAddress(F1:CD:F2:0B:F7:C7,s),)
[2022-05-30 13:58:12,417] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.device.on_driver_event:34] [DEBUG]: Got NRF Driver event: GapEvtConnected(conn_handle=0, peer_addr=BLEGapAddr(F1:CD:F2:0B:F7:C7,s), role=<BLEGapRoles.central: 2>, conn_params=BLEGapConnParams(interval: [30.0-30.0] ms, timeout: 4000.0 ms, latency: 0))
2022-05-30 13:58:12,418 [    INFO] [140409226518528/MainThread] Connected, conn_handle: 0 (main.py:134)
[2022-05-30 13:58:12,418] [MainThread] [__main__.<module>:134] [INFO]: Connected, conn_handle: 0
2022-05-30 13:58:12,419 [    INFO] [140409226518528/MainThread] Starting pairing procedure ... (main.py:151)
[2022-05-30 13:58:12,419] [MainThread] [__main__.<module>:151] [INFO]: Starting pairing procedure ...
[2022-05-30 13:58:12,419] [MainThread] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_authenticate(0, BLEGapSecParams(bond=True, mitm=True, lesc=False, keypress_noti=False, io_caps=<BLEGapIoCaps.KEYBOARD_ONLY: 2>, oob=False, min_key_size=7, max_key_size=16, kdist_own=BLEGapSecKeyDist(enc_key=True, id_key=False, sign_key=False, link_key=False), kdist_peer=BLEGapSecKeyDist(enc_key=True, id_key=False, sign_key=False, link_key=False)))
[2022-05-30 13:58:12,478] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.device.on_driver_event:34] [DEBUG]: Got NRF Driver event: GapEvtSecParamsRequest(conn_handle=0, sec_params=BLEGapSecParams(bond=1, mitm=1, lesc=0, keypress_noti=0, io_caps=0, oob=0, min_key_size=0, max_key_size=16, kdist_own=BLEGapSecKeyDist(enc_key=1, id_key=0, sign_key=0, link_key=0), kdist_peer=BLEGapSecKeyDist(enc_key=1, id_key=0, sign_key=0, link_key=0)))
[2022-05-30 13:58:12,479] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_sec_params_reply(0, <BLEGapSecStatus.success: 0>, None, BLEGapSecKeyset(own: BLEGapSecKeys(enc: key: , master_id: BLEGapMasterId(e: 0, r: b''), id: , sign: , pk: a6298272a3f05001988e3a50fa4093078920ab592761138a380b894ce4aea13d1376b9ae3f7c2140cf97fb9363c28e51a6e4bd7213bf8324bf095bd63c937ee9), peer: BLEGapSecKeys(enc: key: , master_id: BLEGapMasterId(e: 0, r: b''), id: , sign: , pk: )))
[2022-05-30 13:58:12,479] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.nrf.nrf_driver.wrapper:67] [DEBUG]: time to break
[2022-05-30 13:58:12,488] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.device.on_driver_event:34] [DEBUG]: Got NRF Driver event: GapEvtAuthKeyRequest(conn_handle=0, key_type=<BLEGapAuthKeyType.PASSKEY: 1>)
2022-05-30 13:58:12,488 [    INFO] [140409049179712/0 Passkey Entry] Passkey Requested (main.py:46)
[2022-05-30 13:58:12,488] [0 Passkey Entry] [__main__.on_passkey_entry:46] [INFO]: Passkey Requested
[2022-05-30 13:58:12,488] [0 Passkey Entry] [blatann.nrf.nrf_driver.wrapper:64] [DEBUG]: [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00] ble_gap_auth_key_reply(0, <BLEGapAuthKeyType.PASSKEY: 1>, b'000000')
[2022-05-30 13:58:12,628] [/dev/serial/by-id/usb-SEGGER_J-Link_000682553260-if00_Event] [blatann.device.on_driver_event:34] [DEBUG]: Got NRF Driver event: GapEvtDisconnected(conn_handle=0, reason=<BLEHci.authentication_failure: 5>)
```

Would anyone have any idea where to look at for troubleshooting this problem?

Note that I have managed to successfully pair the same nRF DK (using exactly the same Nordic Connectivity FW) with the same peripheral using Nordics `pc-ble-driver-py` directly. For this all I did was implement an over-ride for the drivers `on_gap_evt_auth_key_request()` as follows

```python

    def on_gap_evt_auth_key_request(self, ble_driver, conn_handle, **kwargs):
        LOGGER.info(f'S/N {self.serial}: Pass Key requested')
        passkey = b'000000'
        from pc_ble_driver_py.ble_driver import (driver, util)
        pk = util.list_to_uint8_array(passkey)

        driver.sd_ble_gap_auth_key_reply(
            ble_driver.rpc_adapter,
            conn_handle,
            kwargs["key_type"],
            pk.cast(),
        )
```
and then trigger pairing via the driver's `authenticate()` method.

Thanks in advance for any support.

**Comment by rene-becker-setec** [comment]
I just realized that a new version 0.4.0 has been released - so far I have been using 0.3.6.
I better give that one a go ...

**Comment by rene-becker-setec** [comment]
All working after upgrading to v0.4.0
